homeassistant/components/heos/__init__.py

```diff
@@ -25,28 +26,59 @@
 
 
 async def async_setup(hass: HomeAssistantType, config: ConfigType):
-    """Set up the HEOS component."""
-    from aioheos import AioHeosController
-
+    """Create config entry from config if not already created."""
     host = config[DOMAIN][CONF_HOST]
-    controller = AioHeosController(hass.loop, host)
+    entries = hass.config_entries.async_entries(DOMAIN)
+    if not entries:
+        # Create new entry based on config
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN, context={'source': 'import'},
+                data={CONF_HOST: host}))
+    else:
+        # Check if host needs to be updated
+        entry = entries[0]
+        if entry.data[CONF_HOST] != host:
+            entry.data[CONF_HOST] = host
+            entry.title = format_title(host)
+            hass.config_entries.async_update_entry(entry)
 
-    try:
-        await asyncio.wait_for(controller.connect(), timeout=5.0)
-    except asyncio.TimeoutError:
-        _LOGGER.error('Timeout during setup.')
-        return False
+    return True
 
-    async def controller_close(event):
-        """Close connection when HASS shutsdown."""
-        await controller.close()
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, controller_close)
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Initialize config entry which represents the HEOS controller."""
+    from pyheos import Heos
+    # Setting all_progress_events=False ensures that we only receive a
+    # media position update upon start of playback or when media changes
+    controller = Heos(entry.data[CONF_HOST], all_progress_events=False)
 
-    hass.data.setdefault(DOMAIN, {})
-    hass.data[DOMAIN][MEDIA_PLAYER_DOMAIN] = controller
+    try:
+        await controller.connect(auto_reconnect=True)
 
-    hass.async_create_task(async_load_platform(
-        hass, MEDIA_PLAYER_DOMAIN, DOMAIN, {}, config))
+        async def disconnect_controller(event):
+            await controller.disconnect()
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                   disconnect_controller)
+        hass.data[DOMAIN] = {
+            DATA_CONTROLLER: controller,
+            MEDIA_PLAYER_DOMAIN: await controller.get_players()
+        }
+    # Auto reconnect only operates if initial connection was successful.
+    except (asyncio.TimeoutError, ConnectionError):
+        await controller.disconnect()
+        _LOGGER.exception("Error connecting and retrieving players")
```

Comment:
If we know the error we don't need to log a stack trace. Use error level.

Reply:
Well I think we want to capture what kind of error it was.  `ConnectionError` is base to `BrokenPipeError`, `ConnectionAbortedError`, `ConnectionRefusedError` and `ConnectionResetError`.  Knowing which one it was would aid in troubleshooting. Would you prefer to see it logged at debug instead?

Reply:
We can include the name of the error in the log message but keep level at error to avoid the stack trace.

Reply:
Yeah also the troubleshooting would help to know if it error at `connect` or  `get_players` though.

Reply:
We can also use separate try...except wrappers for the different calls to gain more detail.

Reply:
Yeah but that creates more code/maintenance... is it that better than leaving the stack trace in?

Reply:
It's good practice to wrap as little as possible in each try...except. I think we should reserve the stack trace for unexpected errors that we need to guard.

Reply:
Done and added a test for this too.